### ISSUE TITLE
Incorrect colorization / Colorization does not revert

### ISSUE THREAD

**Issue by sujayvsarma** [dotnet/roslyn#34930]
**Version Used**: 
Latest stable (2.0.1)

**Steps to Reproduce**:
Example: Function arguments

Define an empty-body function with at least one argument. Arguments are correctly dimmed out indicating they are not being used.

Code snip:
```
void foo(string arg1) {
  // at this point, 'arg1' is dimmed out
}
```

Now start typing any construct, but don't use the argument yet. For example, start typing "if". 

Code snip:
```
void foo(string arg1) {
  if 
  // at this point, arg1 should remain dimmed out, but it lights up
}
```

The argument(s) will light up one by one. If you have more than one argument, the more new statements you type, they will start lighting up. For example, if you have 2 arguments to foo(), arg1 will light up on the first statement/construct and arg2 on the second one. This is even though you have not yet even thought about that argument!

Now go ahead and use the argument in that statement and then edit that line of code to do something else (other than referencing that argument), you will see the argument randomly changes colorization -- sometimes it shows up dimmed out, sometimes it colorizes to say it is being used. 

**Expected Behavior**:
The arguments continue to be in the correct lighted up / colorized state. 

**Actual Behavior**:
But they are not. When you undo the operation, they do not always go back to the previous state.

**Comment by sujayvsarma** [comment]
@jinujoseph FYI/A

**Comment by CyrusNajmabadi** [comment]
> The argument(s) will light up one by one.

It's not really clear what you mean.  can you be more precise about what you are typing and what you are seeing.

Note: i believe the expected behavior of this feature is to be conservative.  i.e. if there are mistakes in your code, or other things it can't figure out, it will assume your symbols are used.  So, if you only type something like:

```c#
void foo(string arg1) {
  if 
  // at this point, arg1 should remain dimmed out, but it lights up
}
```

then your code is in a broken state, and in a broken state we don't want to incorrectly tell you that something is unused when it really may be used.  So we just conservatively show everythign as used.

Tagging @mavasani to see if his design matches what i think is going on here.

**Comment by sujayvsarma** [comment]
@CyrusNajmabadi Am not such a bad coder that I would type "if" and leave it at that. If you read what I said there, I said as soon as you type the "if", arg1 lights up as if it is already been used. That should not happen. 

> The argument(s) will light up one by one
I have already explained that with an example there. Let me try again with a code snip:

```
void foo(string arg1, string arg2, string arg3) {
  if /* as soon as you typed 'if', 'arg1' will light up, now complete the statement however you please */
  for /* as soon as you type this, 'arg2' will light up */
}
```
This is clearly a bug.

**Comment by CyrusNajmabadi** [comment]
> I said as soon as you type the "if", arg1 lights up as if it is already been used. That should not happen.

As i said, i believe that's by design.  Because the analysis is *conservative*. As the code is *broken* we don't want to make assumptions about what is used/unused.  So everything is considered 'used' so that we don't lead users astray by stating they can remove parameters that they actually need.

**Comment by CyrusNajmabadi** [comment]
> I have already explained that with an example there.

Yes, but as i stated: It's not really clear what you mean. can you be more precise about what you are typing and what you are seeing.

> Let me try again with a code snip:

I don't repro this at all.  When i just have the method, with no body, i see:

![image](https://user-images.githubusercontent.com/4564579/55977327-63e49600-5c43-11e9-9e4c-f8ff5a9ea770.png)

Note that all the parameters are shown as unused.

Then, when i add 'if' all of them show as 'used' because of the reasons i listed above:

![image](https://user-images.githubusercontent.com/4564579/55977355-7959c000-5c43-11e9-9828-d3779f3f4e7c.png)

> `/* now complete the statement however you please */

Ok.  I do, now that there are no errors, i get the analysis about unused params:

![image](https://user-images.githubusercontent.com/4564579/55977394-91314400-5c43-11e9-9481-d9c474de30a4.png)

> `for /* as soon as you type this, 'arg2' will light up */

I see this:

![image](https://user-images.githubusercontent.com/4564579/55977426-9ee6c980-5c43-11e9-8313-8c983ac98a49.png)

Which, again, is the behavior i would expect.  Because the code is now broken, it's not clear what is/isn't used in it.

**Comment by sujayvsarma** [comment]
@CyrusNajmabadi What you are seeing **is** what I was seeing. But what I did not know (before this conversation) is that `Because the code is now broken, it's not clear what is/isn't used in it.` and that's why the engine shows it that way.

I **accept** this explanation. I am closing this issue because this is a sufficient "By design / Won't Fix" explanation. :)
